pkg/scheduler/queue/queue.go

```diff
@@ -415,45 +410,72 @@ func (q *RequestQueue) runQuerierOperation(querierID string, operation querierOp
+		// and the Querier connection counts will be decremented as the workers disconnect.
+		return q.queueBroker.notifyQuerierShutdown(querierOp.querierID)
+	case forgetDisconnected:
+		return q.processForgetDisconnectedQueriers()
+	default:
+		panic(fmt.Sprintf("received unknown querier event %v for querier ID %v", querierOp.operation, querierOp.querierID))
+	}
 }
 
+func (q *RequestQueue) processRegisterQuerierConnection(querierID QuerierID) (resharded bool) {
+	q.connectedQuerierWorkers.Inc()
+	return q.queueBroker.addQuerierConnection(querierID)
+}
+
+func (q *RequestQueue) processUnregisterQuerierConnection(querierID QuerierID) (resharded bool) {
+	q.connectedQuerierWorkers.Dec()
+	return q.queueBroker.removeQuerierConnection(querierID, time.Now())
+}
+
+func (q *RequestQueue) processForgetDisconnectedQueriers() (resharded bool) {
+	return q.queueBroker.forgetDisconnectedQueriers(time.Now())
+}
+
+// nextRequestForQuerierCall is a "request" indicating that the querier is ready to receive the next query request.
+// It embeds an unbuffered result channel for the nextRequestForQuerier "response" to be written to.
+// "Request" and "Response" terminology is avoided in the naming here to avoid confusion with the actual query requests.
 type nextRequestForQuerierCall struct {
-	ctx           context.Context
-	querierID     QuerierID
-	lastUserIndex UserIndex
-	processed     chan nextRequestForQuerier
+	ctx             context.Context
+	querierID       QuerierID
+	lastTenantIndex TenantIndex
+	resultChan      chan nextRequestForQuerier
+}
 
-	haveUsed bool // Must be set to true after sending a message to processed, to ensure we only ever try to send one message to processed.
+// nextRequestForQuerier is the "response" to a nextRequestForQuerierCall, to be written to the call's result channel.
+// Errors are embedded in this response rather than written to a separate error channel
+// so that lastTenantIndex can still be returned back to the querier connection.
+type nextRequestForQuerier struct {
+	req             Request
+	lastTenantIndex TenantIndex
+	err             error
 }
 
 func (q *nextRequestForQuerierCall) sendError(err error) {
 	// If GetNextRequestForQuerier is already gone, we don't care, so ignore the result from send.
 	_ = q.send(nextRequestForQuerier{err: err})
 }
 
-// send sends req to the GetNextRequestForQuerier call that is waiting for a new query.
-// Returns true if sending succeeds, or false otherwise (eg. because the GetNextRequestForQuerier call has already returned due to a context
-// cancellation).
+// send sends req to the nextRequestForQuerierCall result channel that is waiting for a new query.
+// Returns true if sending succeeds, or false if req context is timed out or canceled.
 func (q *nextRequestForQuerierCall) send(req nextRequestForQuerier) bool {
-	if q.haveUsed {
-		panic("bug: should not try to send multiple messages to a querier")
-	}
-
-	q.haveUsed = true
```

Comment:
a) closing the `q.processed` (now named `q.resultChan`) channel will also panic if the channel is read from, albeit with the less-descriptive panic message.
b) this has never been logged since this was added
c) it's not possible for it to happen with the current code